server/src/main/kotlin/net/horizonsend/ion/server/features/custom/items/blasters/objects/CratePlacer.kt

```diff
@@ -0,0 +1,266 @@
+package net.horizonsend.ion.server.features.customitems.blasters.objects
+
+
+import com.google.common.collect.ImmutableSet
+import fr.skytasul.guardianbeam.Laser.GuardianLaser
+import net.horizonsend.ion.common.extensions.alert
+import net.horizonsend.ion.server.IonServer
+import net.horizonsend.ion.server.features.custom.items.CustomItem
+import net.horizonsend.ion.server.features.custom.items.CustomItems.customItem
+import net.horizonsend.ion.server.features.custom.items.blasters.objects.AmmunitionHoldingItem
+import net.horizonsend.ion.server.features.custom.items.blasters.objects.Magazine
+import net.horizonsend.ion.server.features.economy.cargotrade.ShipmentManager.getShipmentItemId
+import net.horizonsend.ion.server.miscellaneous.utils.Tasks
+import net.horizonsend.ion.server.miscellaneous.utils.minecraft
+import net.kyori.adventure.key.Key.key
+import net.kyori.adventure.sound.Sound.Source.PLAYER
+import net.kyori.adventure.sound.Sound.sound
+import net.kyori.adventure.text.Component
+import net.kyori.adventure.text.Component.text
+import net.kyori.adventure.text.format.NamedTextColor
+import net.minecraft.core.BlockPos
+import net.minecraft.nbt.StringTag
+import net.minecraft.world.level.block.entity.BlockEntity
+import org.bukkit.ChatColor
+import org.bukkit.FluidCollisionMode
+import org.bukkit.Location
+import org.bukkit.Material
+import org.bukkit.SoundCategory.PLAYERS
+import org.bukkit.block.Block
+import org.bukkit.block.BlockFace
+import org.bukkit.block.ShulkerBox
+import org.bukkit.craftbukkit.v1_20_R3.block.CraftShulkerBox
+import org.bukkit.entity.LivingEntity
+import org.bukkit.entity.Player
+import org.bukkit.event.block.BlockPlaceEvent
+import org.bukkit.inventory.EquipmentSlot
+import org.bukkit.inventory.ItemStack
+import org.bukkit.inventory.meta.BlockStateMeta
+
+abstract class CratePlacer(
+	identifier: String,
+
+	material: Material,
+	customModelData: Int,
+	displayName: Component,
+	val magazineType: Magazine<*>,
+	val soundReloadStart: String,
+	val soundReloadFinish: String,
+) : AmmunitionHoldingItem(identifier, material, customModelData, displayName) {
```

Comment:
Probably best to make this its own thing, not just a copy paste of blaster code, and it probably doesn't need ammo. 

Lots of duplication going on.